cmd/podman/cp.go

```diff
@@ -0,0 +1,366 @@
+	"github.com/containers/libpod/libpod"
+	"github.com/containers/libpod/pkg/chrootuser"
+	"github.com/containers/libpod/pkg/rootless"
+	"github.com/containers/storage"
+	"github.com/containers/storage/pkg/archive"
+	"github.com/containers/storage/pkg/chrootarchive"
+	"github.com/containers/storage/pkg/idtools"
+	digest "github.com/opencontainers/go-digest"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+var (
+	cpDescription = "Copy files/folders between a container and the local filesystem"
+	cpCommand     = cli.Command{
+		Name:         "cp",
+		Usage:        "Copy files/folders between a container and the local filesystem",
+		Description:  cpDescription,
+		Action:       cpCmd,
+		ArgsUsage:    "CONTAINER:SRC_PATH DEST_PATH | SRC_PATH|- CONTAINER:DEST_PAT",
+		OnUsageError: usageErrorHandler,
+	}
+)
+
+func cpCmd(c *cli.Context) error {
+	args := c.Args()
+	if len(args) != 2 {
+		return errors.Errorf("you must provide source path and destination path")
+	}
+	srcPath := args[0]
+	destPath := args[1]
+
+	if isContainerPathFormat(srcPath) && isContainerPathFormat(destPath) {
+		return errors.Errorf("please provide only one container working dorectory")
+	}
+	if !isContainerPathFormat(srcPath) && !isContainerPathFormat(destPath) {
+		return errors.Errorf("you must provade one container ID")
+	}
+
+	if os.Geteuid() != 0 {
+		rootless.SetSkipStorageSetup(true)
+	}
+
+	runtime, err := libpodruntime.GetRuntime(c)
+	if err != nil {
+		return errors.Wrapf(err, "could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	if os.Geteuid() != 0 {
+		if driver := runtime.GetConfig().StorageConfig.GraphDriverName; driver != "vfs" {
+			// Do not allow to mount a graphdriver that is not vfs if we are creating the userns as part
+			// of the mount command.
+			return fmt.Errorf("cannot mount using driver %s in rootless mode", driver)
+		}
+
+		became, ret, err := rootless.BecomeRootInUserNS()
+		if err != nil {
+			return err
+		}
+		if became {
+			os.Exit(ret)
+		}
+	}
+
+	if isContainerPathFormat(destPath) {
+		name := strings.Split(destPath, ":")[0]
+		destination := strings.Split(destPath, ":")[1]
+		err := copyBetweenHostAndContainer(name, srcPath, destination, runtime, true)
+		if err != nil {
+			return errors.Wrapf(err, "error copy from %s to %s", srcPath, destPath)
+		}
+		return nil
+	}
+
+	if isContainerPathFormat(srcPath) {
+		name := strings.Split(srcPath, ":")[0]
+		srcPath = strings.Split(srcPath, ":")[1]
+		destination := destPath
+		err := copyBetweenHostAndContainer(name, srcPath, destination, runtime, false)
+		if err != nil {
+			return errors.Wrapf(err, "error copy from %s to %s", srcPath, destPath)
+		}
+		return nil
+	}
+	return errors.Errorf("invalid parameters: %s, %s", srcPath, destPath)
+}
+
+func copyBetweenHostAndContainer(name, srcPath, destination string, runtime *libpod.Runtime, isFromHostToContainer bool) error {
```

Comment:
Should check that srcPath and destination are len > 0 immediately so we don't try a copy with garbage data